repoaudit/repoaudit/utils.py

```diff
@@ -26,14 +40,223 @@ def hexdigest(self, alg: str) -> str:
         return self.hashers[alg].hexdigest()
 
 
-def output_result(proc_packages: int, errors: List[str]) -> bool:
-    """Output number of packages processed and errors."""
-    click.echo(f"Checked {proc_packages} package(s).")
-    if errors:
-        click.echo(f"{len(errors)} errors found:\n" + ("\n").join(errors))
+class RepoErrors:
+    """A class to keep track of repository errors. Each repository has an entry
+    for each distro in it. If there are no distros (i.e. yum repo) then
+    YUM_DIST or APT_DIST can be used instead."""
+
+    YUM_DIST = "yum_dist"
+    APT_DIST = "apt_dist"
+
+    def __init__(self) -> None:
+        self.errors = dict()
+
+    def add(self, repo: str, dist: Optional[str], error: Optional[str]) -> None:
+        """
+        - If just a repository is specified without a dist and it hasn't been added yet, it
+        will have the state "empty".
+        - If a dist is specified, the repository it is a part of as well as
+        the dist will have the state "ok".
+        - If an error is specified, a dist and repo must also be specified
+        for which the error is a part of. Upon adding an error, the dist and
+        repo it is a part of will change to the state "error".
+        """
+
+        if repo not in self.errors:
+            self.errors[repo] = dict()
+            self.errors[repo]["state"] = "empty"
+
+        # update time
+        self.errors[repo]["time"] = str(datetime.datetime.utcnow())
```

Comment:
I think you could probably drop this code if you used a `defaultdict`:

```python
from collections import defaultdict

self.errors = defaultdict(lambda: {"state": "empty", "time": datetime.datetime.utcnow(), "dists": dict()})
```

Reply:
defaultdict could be nice here but when outputting the dictionary as a json, if the default value has not been explicitly modified then it will not show up in the json output. One workaround I tested was to do 
```python
self.errors[repo] = self.errors[repo]
```
which then initializes the repo in the dictionary but in my opinion looks like unclear code. Any suggestion on a prettier way of using the default dict?

Reply:
One thing I overlooked is that you’re updating the time right? So you could drop that from the defaultdict and set it every time. So keep this line here and that should initialize the entry for repo.

Reply:
I've changed it to a defaultdict.